[goecharger] Initial contribution

SamuelBrucksch opened the pull request:
Forum post for this binding:
https://community.openhab.org/t/go-e-charger-wallbox-binding/

Product:
https://go-e.co/go-echarger-home/

This binding is for reading and configuring the go echarger mobile wallbox. My goal is to implement  PV charging with openhab and this binding, so i can charge as much solar energy as possible. 

Sample sitemap that shows values you can get from wallbox:
![alt text](https://community-openhab-org.s3.dualstack.eu-central-1.amazonaws.com/original/3X/d/5/d5db35b4c787982fbd03036050ab0f73a8701b8a.jpeg)

Binding .jar zipped:
[org.openhab.binding.goecharger-2.5.3-SNAPSHOT.zip](https://github.com/openhab/openhab-addons/files/4301637/org.openhab.binding.goecharger-2.5.3-SNAPSHOT.zip)

SamuelBrucksch commented:
Will receive my go-eCharger beginning of next week then i will heavily test this and mark as ready for review.

J-N-K commented on bundles/org.openhab.binding.goecharger/src/main/resources/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,167 @@
+		<state pattern="%d A" readOnly="true"></state>
+	</channel-type>
+	<channel-type id="tmp">
+		<item-type>Number</item-type>
```
```suggestion
		<item-type>Number:Temperature</item-type>
```

J-N-K commented on bundles/org.openhab.binding.goecharger/src/main/resources/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,167 @@
+		<item-type>Number</item-type>
+		<label>Temperature</label>
+		<description>Temperature of the Go-eCharger</description>
+		<state pattern="%d °C" readOnly="true"></state>
```
```suggestion
		<state pattern="%d %unit%" readOnly="true"></state>
```

is this only provided as integer?

SamuelBrucksch replied:
Yes

SamuelBrucksch commented on bundles/org.openhab.binding.goecharger/src/main/resources/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,167 @@
+		<item-type>Number</item-type>
+		<label>Max Ampere</label>
+		<description>Max ampere per phase allowed to use for charging</description>
+		<state pattern="%d A" readOnly="false"></state>
```
Yes, you can set it.

SamuelBrucksch commented on bundles/org.openhab.binding.goecharger/src/main/java/org/openhab/binding/goecharger/internal/handler/GoEChargerHandler.java:
```diff
@@ -0,0 +1,402 @@
+                    for (Channel channel : getThing().getChannels()) {
+                        updateChannel(channel.getUID().getId());
+                    }
+                } catch (Exception e) {
```
guess this was a leftover from copying it somewhere else

SamuelBrucksch commented:
Thanks for your review so far, all that is left is checking that the values are really correct now with UoM and the shared http client. Will do the rest tomorrow.

SamuelBrucksch commented:
@J-N-K thanks for pointing out the problems in my PR, i think i fixed them all. The only thing that did not work well was `@NonNullByDefault`in the gson representation class, so i left it out for now. 

Good hints about the UoM and http clients, thanks.

SamuelBrucksch commented:
Looks like i signed off the merges from main branch as well, i hope thats not a problem...

J-N-K commented on bundles/org.openhab.binding.goecharger/src/main/java/org/openhab/binding/goecharger/internal/handler/GoEChargerHandler.java:
```diff
@@ -0,0 +1,356 @@
+
+            result = gson.fromJson(contentResponse.getContentAsString(), GoEStatusResponse.class);
+            // logger.debug("Response: {}", contentResponse.getContentAsString());
+        } catch (Exception e) {
```
see above

J-N-K commented on bundles/org.openhab.binding.goecharger/src/main/java/org/openhab/binding/goecharger/internal/handler/GoEChargerHandler.java:
```diff
@@ -0,0 +1,356 @@
+    }
+
+    private void sendData(String key, String value) {
+        String errorMsg = null;
```
can be removed

J-N-K commented on bundles/org.openhab.binding.goecharger/src/main/java/org/openhab/binding/goecharger/internal/handler/GoEChargerHandler.java:
```diff
@@ -0,0 +1,356 @@
+        }
+
+        logger.warn("Error in Go-eCharger request: {}", errorMsg);
+        updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.OFFLINE.COMMUNICATION_ERROR, errorMsg);
```
```suggestion
        updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.OFFLINE.COMMUNICATION_ERROR, "no response received on command");
```

J-N-K commented on bundles/org.openhab.binding.goecharger/src/main/java/org/openhab/binding/goecharger/internal/handler/GoEChargerHandler.java:
```diff
@@ -0,0 +1,356 @@
+        goeResponse = getGoEData();
+
+        // Update all channels from the updated GoE data
+        getThing().getChannels().forEach(channel -> updateChannel(channel.getUID().getId()));
```
```suggestion
        getThing().getChannels().map(channel -> channel.getUID().getId()).forEach(id -> updateState(id, getValue(id)));
```

You can the remove the updateState method. If you check null on `goeReponse` (and return if empty) above and add it as parameter to `getValue` (making it `private State getValue(String id, GoEStatusResponse goeResponse)`, you can also remove the field, making the whole thing a bit more thread-safe.

SamuelBrucksch replied:
If i check for null on goeResponse i also need to set null on all channels, which updateChannel already does for me

SamuelBrucksch replied:
goeResponse is also used in RefreshCommand, can i ignore the refresh command? then this could work.

druciak commented on bundles/org.openhab.binding.goecharger/src/main/java/org/openhab/binding/goecharger/internal/handler/GoEChargerHandler.java:
```diff
@@ -0,0 +1,356 @@
+    public void handleCommand(ChannelUID channelUID, Command command) {
+        if (command instanceof RefreshType) {
+            updateChannel(channelUID.getId());
+            return;
```
I might be wrong, but I think the purpose of this command is to retrieve current value for given channel directly from the device, so I would call `refresh()` here.
@J-N-K  WDYT?

SamuelBrucksch replied:
good catch, this also works good together with the review remark a few comments above to change 

```
getThing().getChannels().forEach(channel -> updateChannel(channel.getUID().getId()));
```
to
```
getThing().getChannels().map(channel -> channel.getUID().getId()).forEach(id -> updateState(id, getValue(id)));
```

How often is this refresh command triggered? in the end i already refresh like every 5 seconds or so, if there is an additional refresh inbetween i dont think that gives you more/better data.

druciak commented on bundles/org.openhab.binding.goecharger/src/main/java/org/openhab/binding/goecharger/internal/handler/GoEChargerHandler.java:
```diff
@@ -0,0 +1,356 @@
+    private @Nullable GoEChargerConfiguration config;
+
+    private Gson gson;
+    private @Nullable GoEStatusResponse goeResponse;
```
I don't think you need this. Just pass response data from `GetGoEData` to other methods as a parameter.

cpmeister commented on bundles/org.openhab.binding.goecharger/src/main/java/org/openhab/binding/goecharger/internal/handler/GoEChargerHandler.java:
```diff
@@ -0,0 +1,356 @@
+        if (retryCounter == 1) {
+            logger.warn("Error in getting data from Go-E charger, retrying once");
+            return getGoEData();
+        }
```
Why would you need to retry immediately? Your refreshJob does retries just by virtue that it is periodic.

SamuelBrucksch replied:
The idea is to not set null values immediately but try again in case there was just a short connection issue. So only after the next retry values are set to null/undef. Took this overfrom fronius boinding. I guess the values are a bit more stable like this in shaky wifi connections.

SamuelBrucksch commented on bundles/org.openhab.binding.goecharger/src/main/java/org/openhab/binding/goecharger/internal/api/GoEStatusResponse.java:
```diff
@@ -0,0 +1,199 @@
+ *
+ * @author Samuel Brucksch - Initial contribution
+ */
+public class GoEStatusResponse {
```
I'm not so familiar with gson, i thought it needs setter/getters to work. but it works fine if i remove them

SamuelBrucksch commented on bundles/org.openhab.binding.goecharger/src/main/java/org/openhab/binding/goecharger/internal/api/GoEStatusResponse.java:
```diff
@@ -0,0 +1,199 @@
+  }
+
+  public void setEnergy( Integer[] energy) {
+    this.energy = energy;
```
It is used for voltage, current and power. however its either null and there is no connection or it is always there

SamuelBrucksch commented on bundles/org.openhab.binding.goecharger/src/main/java/org/openhab/binding/goecharger/internal/handler/GoEChargerHandler.java:
```diff
@@ -0,0 +1,356 @@
+    }
+
+    public State getValue(String channelId) {
+        String[] fields = StringUtils.split(channelId, "#");
```
This is a leftover from copying stuff from other bindings, guess i will spend some time and clean up fronius binding after this, where i took most of the code from.

SamuelBrucksch commented on bundles/org.openhab.binding.goecharger/src/main/java/org/openhab/binding/goecharger/internal/handler/GoEChargerHandler.java:
```diff
@@ -0,0 +1,356 @@
+                    return new QuantityType<>((Double) (goeResponse.getEnergy()[9] / 10d),
+                            SmartHomeUnits.WATT);
+                default:
+                    return UnDefType.NULL;
```
In the openhablog i always saw values getting NULL if device is not available, so i thought NULL is the right choice here.

SamuelBrucksch commented on bundles/org.openhab.binding.goecharger/src/main/java/org/openhab/binding/goecharger/internal/handler/GoEChargerHandler.java:
```diff
@@ -0,0 +1,356 @@
+     * @param channelId the id identifying the channel to be updated
+     */
+    protected void updateChannel(String channelId) {
+        if (!isLinked(channelId)) {
```
Also leftover from copying code ;) Guess it really is necessary to clean up some other bindings as well.

SamuelBrucksch commented on bundles/org.openhab.binding.goecharger/README.md:
```diff
@@ -0,0 +1,101 @@
+|-----------|------------------------------------------------------------------------- |----------|
+| ip        | the ip-address of your go-eCharger | yes |
+| refreshInterval  | Interval to read data, default 5 (in seconds) | no |
+
```
http://markdowntable.com/

This one is also quite handy to format tables that already exist

SamuelBrucksch commented:
Thanks for the remarks. I changed my code accordingly.

SamuelBrucksch commented on bundles/org.openhab.binding.goecharger/src/main/java/org/openhab/binding/goecharger/internal/handler/GoEChargerHandler.java:
```diff
@@ -0,0 +1,333 @@
+    }
+
+    public State getValue(String channelId, GoEStatusResponseDTO goeResponse) {
+        if (goeResponse != null) {
```
How to set Null/Undef for each channel then? Does openhab do that automatically when the device goes offline? Or is it best practice to keep the last known values?

SamuelBrucksch commented on bundles/org.openhab.binding.goecharger/src/main/java/org/openhab/binding/goecharger/internal/handler/GoEChargerHandler.java:
```diff
@@ -0,0 +1,333 @@
+    private void refresh() {
+        // Request new GoE data
+        retryCounter = 0;
+        GoEStatusResponseDTO goeResponse = getGoEData();
```
I think in a switch you can ommit default, if there is no reason to have a default.

J-N-K commented on bundles/org.openhab.binding.goecharger/src/main/java/org/openhab/binding/goecharger/internal/handler/GoEChargerHandler.java:
```diff
@@ -0,0 +1,333 @@
+        // Request new GoE data
+        retryCounter = 0;
+        GoEStatusResponseDTO goeResponse = getGoEData();
+
```
```suggestion
if (goeResponse == null) {
    allChannels.forEach(channel -> updateState(channel, UndefType.UNDEF));
} else {
   allChannels.forEach(channel -> updateState(channel, getValue(channel, goeResponse)));
}

```

SamuelBrucksch replied:
Seems to work, will do it like this than. Thanks for your input.

J-N-K commented:
Regarding the codestyle: I use VSCode for my openHAB configuration but IntelliJ IDEA for development., so I can't help you here. I would suggest to run `mvn spotless:apply -pl <path to your binding>` once before we merge. That should fix all codestyle issues.

SamuelBrucksch commented:
I think i implemented/adopted everything that was requested now.

druciak commented on bundles/org.openhab.binding.goecharger/src/main/java/org/openhab/binding/goecharger/internal/handler/GoEChargerHandler.java:
```diff
@@ -0,0 +1,333 @@
+    private final Logger logger = LoggerFactory.getLogger(GoEChargerHandler.class);
+
+    private @Nullable GoEChargerConfiguration config;
+
```
I would use `Collections.emptyList()` here.

druciak commented on bundles/org.openhab.binding.goecharger/src/main/java/org/openhab/binding/goecharger/internal/handler/GoEChargerHandler.java:
```diff
@@ -0,0 +1,333 @@
+            case MAX_AMPERE:
+                key = "amp";
+                value = ((QuantityType<?>)command).intValue() + "";
+                break;
```
Can you be sure the command here will be always of QuantityType type? I would split this switch into two `if` checks: `if (command instanceof QuantityType)` and `if (command instanceof OnOffType)`.

SamuelBrucksch replied:
max amps is always quantity type, yes. How could it be something differentif i check for the ID beforehand?

druciak replied:
The channel is of QuantityType, but I am talking about passed command. I believe you can send whatever command you want using rules. Also, when there is new command introduced, something similar to `REFRESH` command, you will get cast exception here.

SamuelBrucksch replied:
Ok i get it. But then it gets quite complicated, command can be either string, or number or whatever. If you send Text Command "6" or  Number Command 6 it could both work.

J-N-K replied:
No. Commands are strongly typed in openHAB. Checking for `instanceof QuantityType` is enough. Commands send from rules are correctly casted before.

SamuelBrucksch replied:
Ok so i do not have to do anything here? Will quickly check with a custom rule...

SamuelBrucksch replied:
```
java.lang.ClassCastException: org.eclipse.smarthome.core.library.types.DecimalType cannot be cast to org.eclipse.smarthome.core.library.types.QuantityType
```
So i guess i have the wrong type here anyway.

SamuelBrucksch replied:
However if i send a string in the rule instead of an int value it still says decimaltype in the exception, so i guess no need to check for different types.

SamuelBrucksch replied:
Changed type to decimaltype and tested a rule with string and number and both worked.

druciak replied:
This is what I do: check type before casting, because the value comes fron the outside and there is no guarantee it will be of proper type. It happened to my binding (satel) when `REFRESH` command had been introduced - the binding was crashing on this command because there wasn't type checking in one handler. 
But if you feel that this is not needed here, just leave as it is.

druciak commented on bundles/org.openhab.binding.goecharger/src/main/java/org/openhab/binding/goecharger/internal/handler/GoEChargerHandler.java:
```diff
@@ -0,0 +1,333 @@
+    @Override
+    @SuppressWarnings("unchecked")
+    public void handleCommand(ChannelUID channelUID, Command command) {
+        if (command instanceof RefreshType) {
```
I would remove that `if` and check the type for each channel. See my comment below.

druciak commented on bundles/org.openhab.binding.goecharger/src/main/java/org/openhab/binding/goecharger/internal/handler/GoEChargerHandler.java:
```diff
@@ -0,0 +1,333 @@
+            if (retryCounter == 1) {
+                logger.warn("Error in getting data from Go-E charger, retrying once");
+                return getGoEData();
+            }
```
I really don't like this. Just return `null` value in case an error and handle retry in `refresh` method. Also I wouldn't do a retry in case of `InterruptedException`. And remove `retryCounter` completely.

SamuelBrucksch replied:
I guess you are right, its not up to the binding to fix bad wifi

druciak commented on bundles/org.openhab.binding.goecharger/src/main/java/org/openhab/binding/goecharger/internal/handler/GoEChargerHandler.java:
```diff
@@ -0,0 +1,333 @@
+
+        if (result != null) {
+            updateStatus(ThingStatus.ONLINE, ThingStatusDetail.NONE);
+            return result;
```
Imo better would be to move it to `refresh` method.

druciak commented on bundles/org.openhab.binding.goecharger/src/main/java/org/openhab/binding/goecharger/internal/handler/GoEChargerHandler.java:
```diff
@@ -0,0 +1,333 @@
+                logger.warn("Error in getting data from Go-E charger, retrying once");
+                return getGoEData();
+            }
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.OFFLINE.COMMUNICATION_ERROR, "No response received");
```
Imo better would be to move it to `refresh` method.

J-N-K commented on bundles/org.openhab.binding.goecharger/src/main/java/org/openhab/binding/goecharger/internal/handler/GoEChargerHandler.java:
```diff
@@ -0,0 +1,333 @@
+            case MAX_AMPERE:
+                key = "amp";
+                value = ((QuantityType<?>)command).intValue() + "";
+                break;
```
```suggestion
    @Override
    public void handleCommand(ChannelUID channelUID, Command command) {
        String key = null;
        String value = null;
        switch (channelUID.getId()) {
            case MAX_AMPERE:
                key = "amp";
                if (command instanceof DecimalType) {
                    value = String.valueOf(((DecimalType) command).intValue());
                } else if (command instanceof QuantityType<?>) {
                    value = String.valueOf((QuantityType<ElectricalCurrent>) command).toUnit(SmarthomeUnits.AMPERE).intValue());
                }
                break;
```

Something like this would be the correct way to do it. This also makes sure that QuantityType commands have the correct unit. Imagine someone sends a command  `10000 mA` (makes no sense, but you never now). You would probably send something like 10 for 10 amps instead of 10000 (which would happen if you take the intValue before conversion.

SamuelBrucksch replied:
Fine for me, I can do that later. Guess this makes it more stable.

J-N-K replied:
Thanks. I really like working with you.

SamuelBrucksch replied:
I just found how to define UoM in rules here:
https://www.openhab.org/docs/configuration/rules-dsl.html#number-item

So i wrote this code in a rule to test it:
```
var current = 5|A
var mA = current.toUnit("mA")
GoEChargerBindingThingMaxAmpere.sendCommand(mA)
```
This causes an error, however `GoEChargerBindingThingMaxAmpere.sendCommand(mA.intValue)` works, `GoEChargerBindingThingMaxAmpere.sendCommand("5000")` also works and is converted to DecimalType.

So to me it looks like it is not possible to send a QuantityType as command. Or am i doing anything wrong here?

J-N-K replied:
```
var current = 5|A
var mA = current.toUnit("mA")
GoEChargerBindingThingMaxAmpere.sendCommand(mA.toString)
```

SamuelBrucksch replied:
Yeah that one works. It's of type QuantityType. Will upload the changes in a sec.

J-N-K commented on bundles/org.openhab.binding.goecharger/src/main/java/org/openhab/binding/goecharger/internal/handler/GoEChargerHandler.java:
```diff
@@ -0,0 +1,345 @@
+                    .timeout(5, TimeUnit.SECONDS).send();
+
+            result = contentResponse.getContentAsString();
+            // TODO check if value really changed
```
I think you can remove this comment. If you did not receive an error, you can assume that it worked.  And the next refresh will show the updated state.
```suggestion
```

SamuelBrucksch replied:
Guess you are right here...

J-N-K commented on bundles/org.openhab.binding.goecharger/src/main/java/org/openhab/binding/goecharger/internal/handler/GoEChargerHandler.java:
```diff
@@ -0,0 +1,345 @@
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
```
Can be removed since you now check the instance
```suggestion
```

J-N-K commented on bundles/org.openhab.binding.goecharger/src/main/java/org/openhab/binding/goecharger/internal/handler/GoEChargerHandler.java:
```diff
@@ -0,0 +1,345 @@
+        this.httpClient = httpClient;
+    }
+
+    public State getValue(String channelId, GoEStatusResponseDTO goeResponse) {
```
```suggestion
    private State getValue(String channelId, GoEStatusResponseDTO goeResponse) {
```

no need to have this public

SamuelBrucksch commented:
DCO proposes `git rebase HEAD~87 --signoff`because of missing signoffs, but i'm getting rebase errors not related to my change, can anyone help me out here please?

SamuelBrucksch commented:
Guess everything is ready now...

SamuelBrucksch commented:
Thanks again @J-N-K your review was really helpful.

J-N-K commented:
Regarding your re-base questions (which I do not find anymore): Never, ever use "merge" on a feature/bugfix-branch. The risk of messing it all up is very high.

Use rebase instead: 

```
git fetch upstream
git rebase upstream/2.5.x
git push origin +<name of your branch>
```

Assuming `upstream` is the "original" openhab repo and `origin` is your repo). If `2.5.x`-branch of your fork is sync-ed with openhab, you can use `origin` instead of `upstream`.  Not 100% sure what the IDE setup is usally doing, it may be that `origin` is correct there.

SamuelBrucksch commented:
Ok thanks for clarification.

wborn commented on bundles/org.openhab.binding.goecharger/src/main/resources/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,167 @@
+		</channels>
+
+		<config-description>
+			<parameter name="ip" type="text" required="true">
```
Please also add:

`<context>network-address</context>`

See: https://www.eclipse.org/smarthome/documentation/development/bindings/xml-reference.html#supported-contexts

wborn commented on bundles/org.openhab.binding.goecharger/src/main/resources/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,167 @@
+				<label>IP Adress</label>
+				<description>The IP adress of the Go-eCharger</description>
+			</parameter>
+			<parameter name="refreshInterval" type="integer" required="false">
```
You can add a unit for UIs here:

```suggestion
			<parameter name="refreshInterval" type="integer" required="false" unit="s">
```

and also add a unitLabel e.g.:

`<unitLabel>s</unitLabel>`

See: https://www.eclipse.org/smarthome/documentation/development/bindings/xml-reference.html#xml-structure-for-configuration-descriptions

wborn commented on bundles/org.openhab.binding.goecharger/src/main/resources/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,167 @@
+	<channel-type id="pwm">
+		<item-type>Number</item-type>
+		<label>PWM signal status</label>
+		<description>Signal status for PWM signal</description>
```
The description allows for explaining what this abbreviation stands for so people uncommon to it don't have to look it up:

```suggestion
		<description>Pulse-width modulation signal status</description>
```

wborn commented on bundles/pom.xml:
```diff
@@ -87,6 +87,7 @@
     <module>org.openhab.binding.fsinternetradio</module>
     <module>org.openhab.binding.ftpupload</module>
     <module>org.openhab.binding.gardena</module>
+    <module>org.openhab.binding.goecharger</module>
```
Please also add the artifact to the BOM in `bom/openhab-addons/pom.xml`

wborn commented on bundles/org.openhab.binding.goecharger/src/main/resources/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,167 @@
+
+		<config-description>
+			<parameter name="ip" type="text" required="true">
+				<label>IP Adress</label>
```
```suggestion
				<label>IP Address</label>
```

wborn commented on bundles/org.openhab.binding.goecharger/src/main/resources/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,167 @@
+		<config-description>
+			<parameter name="ip" type="text" required="true">
+				<label>IP Adress</label>
+				<description>The IP adress of the Go-eCharger</description>
```
```suggestion
				<description>The IP address of the Go-eCharger</description>
```

wborn commented on bundles/org.openhab.binding.goecharger/src/main/java/org/openhab/binding/goecharger/internal/handler/GoEChargerHandler.java:
```diff
@@ -0,0 +1,344 @@
+            case MAX_AMPERE:
+                return new QuantityType<>(goeResponse.maxChargeAmps, SmartHomeUnits.AMPERE);
+            case PWM_SIGNAL:
+                // TODO more readable string value?
```
Is this TODO still relevant?

SamuelBrucksch replied:
Yes, i guess i will output strings instead. The values represent these states:
1: READY_NO_CAR
2: CHARGING
3: WAITING_FOR_CAR
4: CHARGING_DONE_CAR_CONNECTED

wborn commented on bundles/org.openhab.binding.goecharger/src/main/java/org/openhab/binding/goecharger/internal/handler/GoEChargerHandler.java:
```diff
@@ -0,0 +1,344 @@
+
+    private String getUrl(String type) {
+        String urlStr = type;
+        return urlStr.replace("%IP%", StringUtils.trimToEmpty(config.ip));
```
You can use the `type` parameter straight away instead of first assigning it to `urlStr` or rename it from `type` to `urlStr`.

```suggestion
        return type.replace("%IP%", StringUtils.trimToEmpty(config.ip));
```

SamuelBrucksch replied:
Good catch, i think i did something else before and in the end forget to remove it when i did not need it anymore.

wborn commented on bundles/org.openhab.binding.goecharger/src/main/java/org/openhab/binding/goecharger/internal/handler/GoEChargerHandler.java:
```diff
@@ -0,0 +1,344 @@
+                    .timeout(5, TimeUnit.SECONDS).send();
+
+            result = gson.fromJson(contentResponse.getContentAsString(), GoEStatusResponseDTO.class);
+            // logger.debug("Response: {}", contentResponse.getContentAsString());
```
Please remove commented code or uncomment it and make it conditional `logger.isDebugEnabled()`

SamuelBrucksch replied:
Is the condition really necessary? It will log only with correct set log level, or not?

wborn commented on bundles/org.openhab.binding.goecharger/src/main/java/org/openhab/binding/goecharger/internal/handler/GoEChargerHandler.java:
```diff
@@ -0,0 +1,344 @@
+            ContentResponse contentResponse = httpClient.newRequest(urlStr).method(HttpMethod.POST)
+                    .timeout(5, TimeUnit.SECONDS).send();
+
+            result = contentResponse.getContentAsString();
```
The TODO is still there. Looks like `result` is only used for logging and `getContentAsString` could be relatively expensive. The code would be more compact when you move everything into the try block, e.g.:

```java
        try {
            ContentResponse contentResponse = httpClient.newRequest(urlStr).method(HttpMethod.POST)
                    .timeout(5, TimeUnit.SECONDS).send();
            if (logger.isDebugEnabled()) {
                logger.debug("Response: {}", contentResponse.getContentAsString());
            }
            updateStatus(ThingStatus.ONLINE, ThingStatusDetail.NONE);
        } catch (InterruptedException | TimeoutException | ExecutionException e) {
            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.OFFLINE.COMMUNICATION_ERROR,
                    "No response received on command");
        }
```

SamuelBrucksch replied:
Guess the result is not needed at the moment, might be better to just check the response code. Here i agree with the isDebugEnabled condition.

SamuelBrucksch commented:
Thanks @wborn will upload the required changes later this evening.

wborn commented on bundles/org.openhab.binding.goecharger/src/main/resources/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,169 @@
+	<channel-type id="ast">
+		<item-type>String</item-type>
+		<label>Access State</label>
+		<description>Current access state of the go-eCharge</description>
```
Since all String states are known, you can add state options for these. That makes them translatable and it allows them to show up nicer in UIs:

```xml
		<state readOnly="true">
			<options>
				<option value="OPEN">Open</option>
				<option value="RFID">RFID</option>
				<option value="TIMER">Timer</option>
				<option value="AWATTAR">aWATTar</option>
				<option value="UNKNOWN">Unknown</option>
			</options>
		</state>
```

See also: https://www.openhab.org/docs/developer/bindings/thing-xml.html#state-description

wborn commented on bundles/org.openhab.binding.goecharger/src/main/resources/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,169 @@
+	<channel-type id="pwm">
+		<item-type>String</item-type>
+		<label>PWM signal status</label>
+		<description>Pulse-width modulation signal status</description>
```
Same here:

```xml
		<state readOnly="true">
			<options>
				<option value="READY_NO_CAR">Ready (no car)</option>
				<option value="CHARGING">Charging</option>
				<option value="WAITING_FOR_CAR">Waiting for car</option>
				<option value="CHARGING_DONE_CAR_CONNECTED">Charging done (car connected)</option>
			</options>
		</state>
```

wborn commented on bundles/org.openhab.binding.goecharger/src/main/resources/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,169 @@
+	<channel-type id="err">
+		<item-type>String</item-type>
+		<label>Error Code</label>
+		<description>Error code of Go-eCharger</description>
```
```xml
		<state readOnly="true">
			<options>
				<option value="NONE">None</option>
				<option value="RCCB">RCCB</option>
				<option value="NO_GROUND">No ground</option>
				<option value="INTERNAL">Internal</option>
			</options>
		</state>
```

wborn commented on bundles/org.openhab.binding.goecharger/src/main/resources/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,169 @@
+	<channel-type id="fmw">
+		<item-type>String</item-type>
+		<label>Firmware</label>
+		<description>Firmware Version</description>
```
Seems to be read only so you can add:

```xml
<state readOnly="true" />
```

SamuelBrucksch commented:
Thanks @wborn i added the state fields and also did some minor improvements and renamed accessState to accessConfig and added the command handling of that one as it is writable.

SamuelBrucksch commented:
Awesome. Thanks for the help to all reviewers.
